File: jiwon/code/week14/LCS.py
str1 = input()
str2 = input()

def LCS(str1, str2):
    n = len(str1)
    m = len(str2)

    # dp 테이블 초기화
    dp = [[0] * (m + 1) for _ in range(n + 1)]


    for i in range(1, n + 1):
        for j in range(1, m + 1):
            # 문자가 같을때
            if str1[i-1] == str2[j-1]:
                # 대각선 위의 값 (즉, 이전 위치의 LCS 길이)에 1을 더한 값을 현재 위치에 저장
                dp[i][j] = dp[i - 1][j - 1] + 1
            # 문자가 다를때
            elif str1[i-1] != str2[j-1]:
                # 이전 행 또는 이전 열의 값 중 더 큰 값(즉, 더 긴 LCS)을 현재 위치에 저장
                dp[i][j] = max(dp[i-1][j], dp[i][j-1])

    return dp[n][m]


print(LCS(str1, str2))

File: ilkwon/code/week14/LCS.java
import java.util.*;
import java.io.*;

class LCS {
    static char[] str1;
    static char[] str2;
    static Integer[][] dp;
    static int max = Integer.MIN_VALUE;

    public static void main(String[] argc) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        str1 = br.readLine().toCharArray(); // 문자열 1 입력
        str2 = br.readLine().toCharArray(); // 문자열 2 입력

        dp = new Integer[str1.length][str2.length];


        System.out.println(lcs(str1.length-1, str2.length-1));
    }
    public static int lcs(int a, int b) {

        if(a<0||b<0)
            return 0; //비교 불가 -> 공통문자열이 만들어 질 수 없을때

        if(dp[a][b]==null) { // 결과 계산이 아직되지 않았을 때
            dp[a][b]=0;
            if(str1[a] == str2[b]) // 해당 인덱스에 존재하는 문자가 같다면,
                dp[a][b] = lcs(a-1, b-1)+1; // 두 문자가 LCS에 포함된다 -> dp[a][b]에 저장
            else
                // 두 문자가 다르다면 계속해서 찾는 과정 진행 -> lcs(a-1, b)와 lcs(a, b-1) 중에서 더 큰 값을 선택하여 저장
                dp[a][b] = Math.max(lcs(a-1,b), lcs(a,b-1));
        }

        return dp[a][b];
    }

}